### PR TITLE
image-segmentation plugin 

### PR DESCRIPTION
**Pull request by vhoudebine**
initial commit, macro + plugin recipe using pre-trained maskRCNN model on COCO dataset.
1. Install plugin
2. create codenv
3. run macro to download weights
4. add recipe with 2 input folder : 1 with images to score and 1 with the model weights

### CODE REVIEW

File: image-segmentation/python-runnables/download-weights/runnable.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        output_folder_found = False
+        
+        for folder in project.list_managed_folders():
+            if output_folder_name == folder['name']:
+                output_folder = project.get_managed_folder(folder['id'])
+                output_folder_found = True
+                break
+        
+        if not output_folder_found:
+            output_folder = project.create_managed_folder(output_folder_name)
```
Comment (du-phan) [inline image-segmentation/python-runnables/download-weights/runnable.py]: If the `output_folder_name` was not specified by user, its value will be an empty string -> error ?

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+# Checking whether pycocotools is installed
+try :
+    import pycocotools
+except : 
+    print("##### WARNING ##### Couldn't find pycocotools in installed packages : will proceed to installing the package in plugin codenv")
+    client = dataiku.api_client()
+    codenv = client.get_code_env('python','plugin_image-segmentation_managed')
+    codenv_def = condev.get_definition()
+    codenv_def['specPackageList'] = 'git+https://github.com/waleedka/coco.git#subdirectory=PythonAPI'
+    codenv.set_definition(codenv_def)
```
Comment (du-phan) [inline image-segmentation/custom-recipes/score_image/recipe.py]: to improve: catch error if code-env update fails.

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    
+    print('#'*50)
+    print('Start updating codenv plugin_image-segmentation_managed with git+https://github.com/waleedka/coco.git#subdirectory=PythonAPI')
+    codenv.update_packages()
+    print ('Codenv updated')
+# Import Mask RCNN
+from mrcnn import utils
```
Comment (du-phan) [inline image-segmentation/custom-recipes/score_image/recipe.py]: multiple line import of the same package, should merge them

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    image = skimage.io.imread(image_path)
+    image_list.append(image)
+
+result=model.detect(image_list, verbose=1)
+
+for i, im in enumerate(result):
+    result[i]["image_path"] = to_score[i]
+result_json=[]
+
+def jsonify(data):
```
Comment (du-phan) [inline image-segmentation/custom-recipes/score_image/recipe.py]: to improve: put all classes/functions inside python lib

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        json_data[key] = value
+    return json_data
+
+for res in result:
+    result_json.append(jsonify(res))
+
+with open(DETAILS_DIR + '/results.json', 'w') as outfile:
+    json.dump(result_json, outfile)
+
+output_images = get_recipe_config().get('save_image')
```
Comment (du-phan) [inline image-segmentation/custom-recipes/score_image/recipe.py]: to improve: get all the recipe config at one place, better to follow/update later

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+#############################
+# Your original recipe
+#############################
+
+# -------------------------------------------------------------------------------- NOTEBOOK-CELL: CODE
+# -*- coding: utf-8 -*-
+import dataiku
```
Comment (du-phan) [inline image-segmentation/custom-recipes/score_image/recipe.py]: that's A LOT of import, do you use them all ?

**Comment by vhoudebine** [comment]
Hey @du-phan , thanks for the review. Fixed your points

File: image-segmentation/custom-recipes/score_image/recipe.py (hunk 2 above)
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: added this as per Clement's advice
`with open('../../../../../../config/plugins/image-segmentation/settings.json','r') as f:
        codenv_name = json.load(f).get('codeEnvName')`

File: image-segmentation/custom-recipes/score_image/recipe.py (hunk 3 above)
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: ok done

File: image-segmentation/python-runnables/download-weights/runnable.py (hunk 1 above)
Comment (vhoudebine) [reply image-segmentation/python-runnables/download-weights/runnable.py]: put a Try / except raise ValueError

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+from matplotlib import patches,  lines
+from matplotlib.patches import Polygon
+
+# Checking whether pycocotools is installed
+try :
+    import pycocotools
+except : 
+    print("##### WARNING ##### Couldn't find pycocotools in installed packages : will proceed to installing the package in plugin codenv")
+    client = dataiku.api_client()
+    codenv = client.get_code_env('python','plugin_image-segmentation_managed')
```
[the remark this answers is not in the document]
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: added this as per Clement's advice
with open('../../../../../../config/plugins/image-segmentation/settings.json','r') as f: codenv_name = json.load(f).get('codeEnvName')

File: image-segmentation/custom-recipes/score_image/recipe.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import colorsys
+from skimage.measure import find_contours
+from matplotlib import patches,  lines
+from matplotlib.patches import Polygon
+
+# Checking whether pycocotools is installed
+try :
+    import pycocotools
+except : 
+    print("##### WARNING ##### Couldn't find pycocotools in installed packages : will proceed to installing the package in plugin codenv")
```
[the remark this answers is not in the document]
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: thanks cool lib I didn't know it

File: image-segmentation/custom-recipes/score_image/recipe.py (hunk 2 above)
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: done

File: image-segmentation/custom-recipes/score_image/recipe.py (hunk 6 above)
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: nope, I removed the unnecessary ones

File: image-segmentation/custom-recipes/score_image/recipe.py (hunk 5 above)
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: done

File: image-segmentation/custom-recipes/score_image/recipe.py (hunk 4 above)
Comment (vhoudebine) [reply image-segmentation/custom-recipes/score_image/recipe.py]: done, in utils.py

**Pull request merged**